ptf/tests/linerate/meter_tests.py

```diff
@@ -0,0 +1,654 @@
+        Setup slice level tables and meter (interface table + slice meter).
+        :param slice_rate:
+        :return:
+        """
+        # Slice level configuration
+        self.add_s1u_iface(s1u_addr=N3_ADDR, slice_id=DEFAULT_SLICE_ID)
+        self.add_slice_tc_meter(slice_id=DEFAULT_SLICE_ID, tc=DEFAULT_TC, committed_rate=1, peak_rate=slice_rate)
+
+    def setup_ue_ul(
+            self, ue_addr, ul_teid, app_meter_idx=DEFAULT_APP_METER_IDX, app_meter_bps=None,
+            sess_meter_idx=DEFAULT_SESSION_METER_IDX, session_meter_bps=None, app_id=NO_APP_ID
+    ) -> None:
+        """
+        Setup UE uplink table entries (UL session and UL termination) and eventually
+        the session meter and app meter.
+        :param ue_addr:
+        :param ul_teid:
+        :param app_meter_idx:
+        :param app_meter_bps:
+        :param sess_meter_idx:
+        :param session_meter_bps:
+        :param app_id:
+        :return:
+        """
+        if sess_meter_idx != DEFAULT_SESSION_METER_IDX and session_meter_bps is not None:
+            self.add_qer_session_meter(sess_meter_idx, session_meter_bps)
+        if app_meter_idx != DEFAULT_APP_METER_IDX and app_meter_bps is not None:
+            self.add_qer_app_meter(app_meter_idx, app_meter_bps)
+        self.setup_uplink_ue_session(
+            teid=ul_teid, tunnel_dst_addr=N3_ADDR, session_meter_idx=sess_meter_idx
+        )
+        self.setup_uplink_termination(
+            ue_session=ue_addr, ctr_id=UPF_CTR_IDX, tc=DEFAULT_TC, app_id=app_id,
+            app_meter_idx=app_meter_idx
+        )
+
+    def setup_acl_forwarding(self, in_ports, out_port) -> None:
+        """
+        Setup ACL table to perform forwarding based on input ports.
+        :param in_ports:
+        :param out_port:
+        :return:
+        """
+        # Setup ingress port VLAN table to let traffic into the pipeline
+        for port in set(in_ports + [out_port]):
+            self.setup_port(port, DEFAULT_VLAN, PORT_TYPE_EDGE)
+        # Do actual forwarding via ACL
+        for in_port in in_ports:
+            self.add_forwarding_acl_set_output_port(out_port, ig_port=in_port)
+
+    # Create a stream with GTP encapped traffic.
+    def create_gtp_stream(
+            self, ue_addr, teid, pg_id=None, dport=None, l2_size=None, l1_bps=None,
+    ) -> STLStream:
+        if dport is not None:
+            pkt = simple_udp_packet(ip_src=ue_addr, pktlen=l2_size - GTPU_HDR_BYTES, udp_dport=dport)
+        else:
+            pkt = simple_udp_packet(ip_src=ue_addr, pktlen=l2_size - GTPU_HDR_BYTES)
+        pkt = pkt_add_gtp(pkt, out_ipv4_src=ENB_IPV4, out_ipv4_dst=N3_ADDR, teid=teid)
+        stats = None
+        if pg_id is not None:
+            stats = STLFlowLatencyStats(pg_id=pg_id)
+        return STLStream(
+            packet=STLPktBuilder(pkt=pkt),
+            mode=STLTXCont(bps_L1=l1_bps),
+            flow_stats=stats,
+        )
+
+    def average_monitored_port_stats(self, stats):
```

Comment:
Is the average providing enough accuracy? If the goal is to make sure the traffic generation rate was constant, would it be better to look at the maximum/minimum instead?

Reply:
I'm interested in both sent and received traffic, port stats coming from TRex at the end of the test are not sufficiently accurate. So, I thought a better approximation was the average from live stats.